### ISSUE TITLE
When using `save_pretrained` to save a model loaded with `from_pretrained`, its size becomes twice as large.

### ISSUE THREAD

**Issue by CLL112** [huggingface/transformers#35072]
### System Info

- `transformers` version: 4.46.1
- Platform: Linux-5.15.0-25-generic-x86_64-with-glibc2.35
- Python version: 3.10.15
- Huggingface_hub version: 0.26.2
- Safetensors version: 0.4.5
- Accelerate version: 1.0.1
- Accelerate config:    - compute_environment: LOCAL_MACHINE
        - distributed_type: DEEPSPEED
        - mixed_precision: fp16
        - use_cpu: False
        - debug: False
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - deepspeed_config: {'gradient_accumulation_steps': 1, 'gradient_clipping': 1.0, 'offload_optimizer_device': 'none', 'offload_param_device': 'none', 'zero3_init_flag': False, 'zero_stage': 2}
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
- PyTorch version (GPU?): 2.5.1+cu124 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA GeForce RTX 3090

### Who can help?

@ArthurZucker @Rocketknight1  @muellerzr

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
import transformers
from transformers import AutoTokenizer

model = transformers.OPTForCausalLM.from_pretrained("facebook/opt-6.7b")
tokenizer = AutoTokenizer.from_pretrained("facebook/opt-6.7b")

model.save_pretrained("/data/llm/opt-6.7b-testSave")
tokenizer.save_pretrained("/data/llm/opt-6.7b-testSave")
```

### Expected behavior

The weights of `facebook/opt-6.7b` downloaded from Hugging Face are 12.4GB. After using `save_pretrained`, the saved weights became 24.8GB, which is twice as large as the original size. This leads to an "OUT OF MEMORY" error when I load the model for inference. I want the model saved with `save_pretrained` to be the same size as the downloaded one.

**Comment by ArthurZucker** [comment]
Hey! This is expected: if you check the dtype of the model it is in `float32` because you did not specify `torch_dtype=torch.float16` or `torch_dtype="auto"` will be fixed by #34919

**Comment by CLL112** [comment]
Got it, I understand now. Thank you.

> Hey! This is expected: if you check the dtype of the model it is in `float32` because you did not specify `torch_dtype=torch.float16` or `torch_dtype="auto"` will be fixed by #34919
